src/main/java/com/google/api/gax/grpc/ProviderManager.java

```diff
@@ -0,0 +1,155 @@
+import java.util.concurrent.ScheduledExecutorService;
+
+/**
+ * ProviderManager gives a way to manage the lazy creation of channels and executors that
+ * are shared among multiple API wrappers.
+ *
+ * <pre>
+ * <code>
+ *   ProviderManager providerManager =
+ *     ProviderManager.newBuilder()
+ *       .setChannelProvider(SomeApiSettings.defaultChannelProviderBuilder().build())
+ *       .setExecutorProvider(SomeApiSettings.defaultExecutorProviderBuilder().build())
+ *       .build();
+ *
+ *   SomeApiSettings settingsA =
+ *     SomeApiSettings.defaultBuilder()
+ *       .setExecutorProvider(providerManager)
+ *       .setChannelProvider(providerManager)
+ *       .build();
+ *   SomeApiSettings settingsB =
+ *     SomeApiSettings.defaultBuilder()
+ *       .setExecutorProvider(providerManager)
+ *       .setChannelProvider(providerManager)
+ *       .build();
+ *
+ *   providerManager.shutdown();
+ * </code>
+ * </pre>
+ */
+public class ProviderManager implements ChannelProvider, ExecutorProvider {
+  private final InstantiatingChannelProvider channelProvider;
+  private final InstantiatingExecutorProvider executorProvider;
+  private ManagedChannel channel;
+  private ScheduledExecutorService executor;
+
+  private ProviderManager(
+      InstantiatingChannelProvider channelProvider,
+      InstantiatingExecutorProvider executorProvider) {
+    this.channelProvider = channelProvider;
+    this.executorProvider = executorProvider;
+  }
+
+  @Override
+  public boolean shouldAutoClose() {
+    return false;
+  }
+
+  @Override
+  public ScheduledExecutorService getExecutor() {
+    if (executor == null) {
+      executor = executorProvider.getExecutor();
+    }
+    return executor;
+  }
+
+  @Override
+  public boolean needsExecutor() {
+    return false;
+  }
+
+  @Override
+  public ManagedChannel getChannel() throws IOException {
+    if (channel == null) {
+      if (channelProvider.needsExecutor()) {
+        channel = channelProvider.getChannel(getExecutor());
+      } else {
+        channel = channelProvider.getChannel();
+      }
+    }
+    return channel;
+  }
+
+  @Override
+  public ManagedChannel getChannel(Executor executor) {
+    throw new IllegalStateException("getChannel(Executor) called when needsExecutor() is false");
+  }
+
+  public void shutdown() {
+    if (channel != null) {
+      channel.shutdown();
+    }
+    if (executor != null) {
+      executor.shutdown();
+    }
+  }
+
+  public void shutdownNow() {
+    if (channel != null) {
+      channel.shutdownNow();
+    }
+    if (executor != null) {
+      executor.shutdownNow();
+    }
+  }
+
+  public static Builder newBuilder() {
+    return new Builder();
+  }
+
+  public static class Builder {
+    private InstantiatingChannelProvider channelProvider;
+    private InstantiatingExecutorProvider executorProvider;
+
+    private Builder() {}
+
+    public Builder setChannelProvider(InstantiatingChannelProvider channelProvider) {
```

Comment:
> But would it maybe be clearer to just have ProviderManager accept any ExecutorProvider instead of an InstantiatingExecutorProvider, so that you can have everything go through ProviderManager?

This would probably be nicer, though slightly asymmetric.